src/backend/pipeline/cont_analyze.c

```diff
@@ -1886,20 +1890,58 @@ truncate_timestamp_field(Node *expr, ContAnalyzeContext *context)
 	if (!truncated)
 	{
 		FuncCall *func = makeNode(FuncCall);
-		func->funcname = list_make1(makeString(DATE_TRUNC_SECOND));
-		func->args = list_make1(copyObject(expr));
-		expr = (Node *) func;
+		func->args = list_make1(copyObject(time));
+
+		/* For sliding windows we try to keep around a certain "fan in" threshold */
+		if (sw_expr)
+		{
+			A_Expr *ct_expr;
+			Expr *expr;
+			Const *c;
+			Interval *i;
+			char *fname;
+
+			if (equal(sw_expr->lexpr, time))
+				ct_expr = (A_Expr *) sw_expr->rexpr;
+			else
+				ct_expr = (A_Expr *) sw_expr->lexpr;
+
+			expr = (Expr *) transformExpr(make_parsestate(NULL), ct_expr->rexpr, EXPR_KIND_WHERE);
+			Assert(IsA(expr, Const));
+			c = (Const *) expr;
+			Assert(c->consttype == INTERVALOID);
+
+			i = (Interval *) c->constvalue;
+
+			/* We make the step size one unit smaller than the granularity of the window size unit */
+			if (i->month >= 12)
```

Comment:
I think days should be the coarsest step size when automatically choosing. It might even make sense to make it hours. Only seeing updates once per month or day seems like the other extreme of the scale in terms of intuitive behavior that users expect.